### PR TITLE
Skip Pulumi CLI download if runner version satisfies requested CLI version

### LINKED ISSUE #789: Optimization: Skip Download Pulumi CLI when already present
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

With the exception of MacOS and self-hosted runners, Pulumi comes installed on GitHub Actions runners by default. Right now, when a user does not specify a specific version of Pulumi to use, this action defaults to `latest`. Instead, we should use whatever version of Pulumi is installed on the runner, if one exists. This will remove the need to re-download Pulumi for the vast majority of users (anyone who is using Windows or Ubuntu runners).

As described in https://github.com/pulumi/actions/issues/787, this would eliminate `~100MBs` of network traffic. And, detailed in https://github.com/pulumi/actions/issues/674, it would allow the action to be used in the context of where outbound traffic is blocked.

### Affected area/feature

[The `downloadCli` function.](https://github.com/pulumi/actions/blob/master/src/libs/pulumi-cli.ts#L34)

### PR DESCRIPTION
**Pull request by ehafenmaier**
- Add functionality to check GitHub runner for the existence of the Pulumi CLI. 
- If the CLI is already installed on the runner check if the runner CLI version matches the requested version from the action input.
- If runner CLI version matches, skip downloading CLI.
- If CLI not installed on runner or runner CLI does not match requested version, proceed to download CLI

```mermaid
flowchart LR
    A(Action Start) -- Check Runner CLI --> B{CLI Installed\n On Runner?}
    B -- Yes --> C{Runner CLI\n Version Match?}
    C -- Yes --> D(Skip CLI Download)
    C -- No --> E(Download CLI)
    B -- No --> E
```

> NOTE: Skipping the CLI download will also skip the outbound network call in the `getVersionObject()` method, which was mentioned in #674 

This resolves #789

### CODE REVIEW

File: src/libs/pulumi-cli.ts [hunk 1]
```diff
@@ -42,6 +61,23 @@ export async function downloadCli(range: string): Promise<void> {
...
 
+  // Check for version of Pulumi CLI installed on the runner
+  const runnerVersion = await getVersion();
+
+  if (runnerVersion) {
+    core.info(`Runner version: ${runnerVersion}`);
+    // Check if runner version matches
+    if (semver.satisfies(runnerVersion, range)) {
+      // If runner version matches, skip downloading CLI by exiting the function
+      core.info(`Runner version ${runnerVersion} matched. Skipping Pulumi CLI download`);
```
Comment (RobbieMcKinstry) [inline src/libs/pulumi-cli.ts:72]:
```suggestion
      core.info(`Pulumi version ${runnerVersion} is already installed on this machine. Skipping downloading.`);
```

File: src/libs/pulumi-cli.ts [hunk 2]
```diff
@@ -42,6 +61,23 @@ export async function downloadCli(range: string): Promise<void> {
...
+      return;
+    } else {
+      core.info('Pulumi CLI on the runner does not match. Proceeding to download');
```
Comment (RobbieMcKinstry) [inline src/libs/pulumi-cli.ts:75]:
```suggestion
      core.info('Pulumi ${runnerVersion} does not satisfy the desired version ${range}. Proceeding to download');
```

File: src/libs/pulumi-cli.ts [hunk 3]
```diff
@@ -42,6 +61,23 @@ export async function downloadCli(range: string): Promise<void> {
...
+    }
+  } else {
+    core.info('Pulumi CLI not installed on the runner. Proceeding to download');
```
Comment (RobbieMcKinstry) [inline src/libs/pulumi-cli.ts:78]:
```suggestion
    core.info('Pulumi is not detected in the PATH. Proceeding to download');
```

File: src/libs/pulumi-cli.ts [hunk 4]
```diff
@@ -12,6 +12,25 @@ export async function isAvailable(): Promise<boolean> {
...
+  core.debug(`StdErr: ${res.stderr}`)
+
+  // Only check for success and if the [stdout] starts with the version
+  // prefix 'v'. If success is true and the runner version is not the 
+  // latest version then the "warning of newer version" will be in [stderr] field, 
+  // which will trigger an else condition if we also check for [stderr === '']
+  if (res.success && res.stdout.startsWith('v'))
+    return res.stdout.substring(1); // Return version without 'v' prefix
+  else
+    return undefined;
```
Comment (RobbieMcKinstry) [inline src/libs/pulumi-cli.ts:31]:
```suggestion
  if (res.success && res.stdout.startsWith('v')) {
    // Return version without 'v' prefix
    return res.stdout.substring(1);
  } else {
    return null;
  }
```

File: src/libs/pulumi-cli.ts [hunk 5]
```diff
@@ -12,6 +12,25 @@ export async function isAvailable(): Promise<boolean> {
   return res.stderr != '' && !res.success ? false : res.success;
 }
 
+export async function getVersion(): Promise<string | undefined> {
```
Comment (RobbieMcKinstry) [inline src/libs/pulumi-cli.ts:15]: My internal style-guide is that undefined should be reserved for instances where a variable is not in scope, while null is intended to convey the absence of a defined value.

That's my preference, but other philosophies might be better informed than mine!

```suggestion
export async function getVersion(): Promise<string | null> {
```

File: src/libs/pulumi-cli.ts [hunk 6]
```diff
@@ -12,6 +12,25 @@ export async function isAvailable(): Promise<boolean> {
...
+  core.debug('Executing command \'pulumi version\' to check version of CLI on the runner');
+  
+  const res = await exec.exec('pulumi', ['version'], true);
+
+  core.debug(`Success: ${res.success}`)
+  core.debug(`StdOut: ${res.stdout}`)
+  core.debug(`StdErr: ${res.stderr}`)
```
Comment (RobbieMcKinstry) [inline src/libs/pulumi-cli.ts:22]: Let's either drop these log lines, or provide a little more context about where they're running so they can be more easily differentiated from similar lines during debugging. If they were useful to you while debugging, I imagine they'd be useful to us too.

File: src/libs/pulumi-cli.ts [hunk 7]
```diff
@@ -42,6 +61,23 @@ export async function downloadCli(range: string): Promise<void> {
 
   core.info(`Configured range: ${range}`);
 
+  // Check for version of Pulumi CLI installed on the runner
+  const runnerVersion = await getVersion();
+
+  if (runnerVersion) {
+    core.info(`Runner version: ${runnerVersion}`);
```
Comment (RobbieMcKinstry) [inline src/libs/pulumi-cli.ts:68]:
```suggestion
```

**Review by RobbieMcKinstry — CHANGES_REQUESTED** [review]
Hey @ehafenmaier another great submission! Thank you very much for your diligent contribution!

I have a few nit-picks around the log text but otherwise it looks great! 
Lastly, would you add an entry to the changelog describing this feature? After that, we can merge :)

File: src/libs/pulumi-cli.ts (hunk 6 above)
Comment (ehafenmaier) [reply src/libs/pulumi-cli.ts:22]: Even though they did help me debug the `ExecResult` and figure out what to use in the `if` statement, I'm fine with removing them. I started with your suggestion of checking `stderr === ''` but discovered that didn't work using the debug logs. I made sure to add a comment block explaining that so I think the debug logs have served their purpose and are not needed anymore.

File: src/libs/pulumi-cli.ts (hunk 5 above)
Comment (dixler) [reply src/libs/pulumi-cli.ts]: Internally, the service team follows https://github.com/microsoft/TypeScript/wiki/Coding-guidelines

https://github.com/microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined

undefined is preferred over null

Comment (RobbieMcKinstry) [reply src/libs/pulumi-cli.ts]: Well, judging by the H1s at the top of the document, I'm inclined to disregard advice without any reasoning provided ;D

Comment (Frassle) [reply src/libs/pulumi-cli.ts]: undefined is the norm for typescript here. The optional type `string?` would expand out to `string | undefined`. The TS tooling probably understands that style better.

Comment (ehafenmaier) [reply src/libs/pulumi-cli.ts]: I fully admit that TypeScript is not my strongest language so I let the tooling guide me.  I'm coming from a C# background so I started with a function return type of `Promise<string?>`. The TS compiler complained with the following message:
```
// JSDoc types can only be used inside documentation comments.(8020)
```
During my investigation into this message I ran across this [blog post](https://paulsebastian.codes/why-isnt-string-allowed-as-a-function-return-value-in-typescript) which laid out a pretty good explanation on why `string?` is considered an invalid return type.

I'm happy to follow whatever style guide is decided upon.

**Pull request closed without merging**